File: sdk/engine/src/toolloop/loop.ts
import type { Message, Tool, EngineEvent, ConfirmFunc, Usage } from '../types.js';
import type { LLMClientInterface } from '../llm/client.js';
import { isDangerous } from './danger.js';
import { truncateMessages } from './truncate.js';

const MAX_ITERATIONS = 25;
const MAX_TOOL_RESULT_LENGTH = 4000;

export interface Executor {
  execute(toolName: string, argsJSON: string, authToken: string): Promise<string>;
}

export class ToolLoop {
  private client: LLMClientInterface;
  private executor: Executor;
  private confirm: ConfirmFunc | null = null;
  private maxCtxTokens = 0;

  constructor(client: LLMClientInterface, executor: Executor) {
    this.client = client;
    this.executor = executor;
  }

  setConfirm(fn: ConfirmFunc | null) { this.confirm = fn; }
  setMaxContextTokens(n: number) { this.maxCtxTokens = n; }

  async run(
    messages: Message[],
    tools: Tool[],
    authToken: string,
    onEvent: (event: EngineEvent) => void,
    signal?: AbortSignal,
  ): Promise<Message[]> {
    const totalUsage = { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0 };

    for (let i = 0; i < MAX_ITERATIONS; i++) {
      signal?.throwIfAborted();

      const truncated = truncateMessages(messages, this.maxCtxTokens);
      const { message: msg, usage } = await this.client.chatStreamWithTools(
        truncated, tools,
        (delta) => onEvent({ type: 'content_delta', data: { delta } }),
        signal,
      );

      if (usage) {
        totalUsage.prompt_tokens += usage.prompt_tokens;
        totalUsage.completion_tokens += usage.completion_tokens;
        totalUsage.total_tokens += usage.total_tokens;
      }

      messages = [...messages, msg];

      // No tool calls — final response
      if (!msg.tool_calls?.length) {
        onEvent({ type: 'content', data: { text: msg.content } });
        this.emitUsage(onEvent, totalUsage);
        return messages;
      }

      // Execute each tool call
      for (const tc of msg.tool_calls) {
        onEvent({ type: 'tool_call', data: { name: tc.function.name, arguments: tc.function.arguments } });

        // Confirmation gate
        if (this.confirm && isDangerous(tc.function.name, tc.function.arguments)) {
          const approved = await this.confirm(tc.function.name, tc.function.arguments);
          if (!approved) {
            const result = 'Operation cancelled by user';
            onEvent({ type: 'tool_result', data: { name: tc.function.name, result } });
            messages = [...messages, { role: 'tool', content: result, tool_call_id: tc.id }];
            continue;
          }
        }

        let result: string;
        try {
          result = await this.executor.execute(tc.function.name, tc.function.arguments, authToken);
        } catch (err) {
          result = `Error: ${err instanceof Error ? err.message : String(err)}`;
        }

        if (result.length > MAX_TOOL_RESULT_LENGTH) {
          result = result.slice(0, MAX_TOOL_RESULT_LENGTH) + '\n...(truncated)';
        }

        onEvent({ type: 'tool_result', data: { name: tc.function.name, result } });
        messages = [...messages, { role: 'tool', content: result, tool_call_id: tc.id }];
      }
    }

    this.emitUsage(onEvent, totalUsage);
    throw new Error(`Max iterations (${MAX_ITERATIONS}) reached`);
  }

  private emitUsage(onEvent: (event: EngineEvent) => void, usage: Usage) {
    if (usage.total_tokens > 0) {
      onEvent({ type: 'usage', data: usage });
    }
  }
}

File: core/toolloop/loop.go
package toolloop

import (
	"context"
	"fmt"
	"strings"

	"github.com/ZacharyZcR/Kelper/core/llm"
)

const MaxIterations = 25

type Executor interface {
	Execute(ctx context.Context, toolName, argsJSON, authToken string) (string, error)
}

// ConfirmFunc is called before executing a tool that looks dangerous.
// Return true to proceed, false to skip.
type ConfirmFunc func(toolName, argsJSON string) bool

type Event struct {
	Type string      `json:"type"`
	Data interface{} `json:"data"`
}

type ToolCallEvent struct {
	Name      string `json:"name"`
	Arguments string `json:"arguments"`
}

type ToolResultEvent struct {
	Name   string `json:"name"`
	Result string `json:"result"`
}

type ContentEvent struct {
	Text string `json:"text"`
}

type ContentDeltaEvent struct {
	Delta string `json:"delta"`
}

type Loop struct {
	client       *llm.Client
	executor     Executor
	confirm      ConfirmFunc
	maxCtxTokens int
}

func New(client *llm.Client, executor Executor) *Loop {
	return &Loop{client: client, executor: executor}
}

func (l *Loop) SetConfirm(fn ConfirmFunc) {
	l.confirm = fn
}

func (l *Loop) SetMaxContextTokens(n int) {
	l.maxCtxTokens = n
}

var dangerousPatterns = []string{
	"delete", "remove", "destroy", "drop", "purge", "reset",
}

func isDangerous(toolName, argsJSON string) bool {
	lower := strings.ToLower(toolName)
	for _, p := range dangerousPatterns {
		if strings.Contains(lower, p) {
			return true
		}
	}
	lowerArgs := strings.ToLower(argsJSON)
	for _, m := range []string{`"delete"`, `"put"`, `"patch"`} {
		if strings.Contains(lowerArgs, m) {
			return true
		}
	}
	return false
}

func (l *Loop) Run(ctx context.Context, messages []llm.Message, tools []llm.Tool, authToken string, onEvent func(Event)) ([]llm.Message, error) {
	for i := 0; i < MaxIterations; i++ {
		truncated := truncateMessages(messages, l.maxCtxTokens)
		msg, err := l.client.ChatStreamWithTools(ctx, truncated, tools, func(delta string) {
			onEvent(Event{Type: "content_delta", Data: ContentDeltaEvent{Delta: delta}})
		})
		if err != nil {
			return messages, fmt.Errorf("LLM call failed: %w", err)
		}

		messages = append(messages, *msg)

		if len(msg.ToolCalls) == 0 {
			onEvent(Event{Type: "content", Data: ContentEvent{Text: msg.Content}})
			return messages, nil
		}

		for _, tc := range msg.ToolCalls {
			onEvent(Event{Type: "tool_call", Data: ToolCallEvent{
				Name:      tc.Function.Name,
				Arguments: tc.Function.Arguments,
			}})

			// Confirmation gate for dangerous operations
			if l.confirm != nil && isDangerous(tc.Function.Name, tc.Function.Arguments) {
				if !l.confirm(tc.Function.Name, tc.Function.Arguments) {
					result := "Operation cancelled by user"
					onEvent(Event{Type: "tool_result", Data: ToolResultEvent{
						Name:   tc.Function.Name,
						Result: result,
					}})
					messages = append(messages, llm.Message{
						Role:       "tool",
						Content:    result,
						ToolCallID: tc.ID,
					})
					continue
				}
			}

			result, err := l.executor.Execute(ctx, tc.Function.Name, tc.Function.Arguments, authToken)
			if err != nil {
				result = fmt.Sprintf("Error: %s", err.Error())
			}
			if len(result) > 4000 {
				result = result[:4000] + "\n...(truncated)"
			}

			onEvent(Event{Type: "tool_result", Data: ToolResultEvent{
				Name:   tc.Function.Name,
				Result: result,
			}})

			messages = append(messages, llm.Message{
				Role:       "tool",
				Content:    result,
				ToolCallID: tc.ID,
			})
		}
	}

	return messages, fmt.Errorf("max iterations (%d) reached", MaxIterations)
}
